lib/tokens.js

```diff
@@ -0,0 +1,687 @@
+ * combination of an external resolving party (the "requester") and a unique
+ * entity. An entity is determined to be unique internally by it having its
+ * own `internalId`. This pairwise resolution enables outside parties to
+ * have limited correlation capability: it may only correlate entities across
+ * all of the tokens it is authorized to resolve.
+ *
+ * Given the above constraints, an attempt to resolve a token may fail if it
+ * has already been resolved by a different requester (or if the token is
+ * invalid or expired). If the token has been resolved by the same requester,
+ * the same pairwise identifier that was returned during the previous
+ * resolution will be returned again.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.requester - The a string that unambiguously
+ *   identifies the party requesting token resolution.
+ * @param {string} options.token - The token to resolve.
+ *
+ * @returns {object} An object containing the `pairwiseToken`.
+ */
+export async function resolve({requester, token}) {
+  // parse token
+  const {batchId, index} = await _parseToken({token});
+
+  while(true) {
+    // get batch document
+    const {tokenBatch} = await _getBatch({id: batchId});
+    const {internalId} = tokenBatch;
+
+    // prepare to do pairwise resolution...
+    const {resolvedList: encoded, resolution = {}} = tokenBatch;
+
+    // parse resolved bitstring
+    const resolvedList = new Bitstring({
+      buffer: await Bitstring.decodeBits({encoded})
+    });
+
+    // find resolution entry for `requester`
+    const encodedRequester = base64url.encode(requester);
+    let entry = resolution[encodedRequester];
+
+    // see if token is already resolved
+    if(resolvedList.get(index)) {
+      // token already resolved, see if requester matches
+      if(entry) {
+        const bs = new Bitstring({
+          buffer: await Bitstring.decodeBits({encoded: entry.list})
+        });
+        if(bs.get(index)) {
+          // token resolved for same requester, return pairwise token
+          const record = await _getPairwiseToken({internalId, requester});
+          const {pairwiseToken: {value: pairwiseToken}} = record;
+          return {pairwiseToken};
+        }
+      }
+      // token already resolved to another requester, can only be
+      // scope-resolved once
+      throw new BedrockError(
+        'Token already used.',
+        'NotAllowedError', {
+          public: true,
+          httpStatusCode: 400
+        });
+    }
+
+    // token is not yet resolved, attempt to resolve it for `requester`...
+
+    // create a pairwise token for the requester if one does not exist
+    const {pairwiseToken: {value: pairwiseToken}} = await _upsertPairwiseToken(
+      {internalId, requester});
+
+    // update requester's resolution info for the token batch
+    let bs;
+    if(entry) {
+      bs = new Bitstring({
+        buffer: await Bitstring.decodeBits({encoded: entry.list})
+      });
```

Comment:
Is this just doing a copy?  Perhaps Bitstring constructor should be able to initialize itself with another Bitstring?

Reply:
I'm probably misreading what's happening here.  But seems like this pattern is elsewhere and maybe there's an improvement to Bitstring that could simplify it.

Reply:
The decoding is asynchronous -- so it can't easily be done in a constructor.

Reply:
Right.  Maybe make an async Bitstring.from(...) helper that uses internal APIs for split construction/initialization.